Implementation for `UnsafeAccessorTypeAttribute`

AaronRobinsonMSFT opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/90081

Example:

Assembly `A`
```csharp
// Note class is private
class Class1
{
    Class1() { }
    List<Class2> GetListOfClass2() => new List<Class2>();
}

// Note class is private
class Class2 { }
```

Assembly `B` - no metadata reference to assembly `A`
```csharp
object class1 = CreateClass();
object listClass2 = CallGetListOfClass2(class1);
Assert.Equal("System.Collections.Generic.List`1[[Class2, A, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]]", listClass2.GetType().FullName);

[UnsafeAccessor(UnsafeAccessorKind.Constructor)]
[return: UnsafeAccessorType("Class1, A")]
extern static object CreateClass();

[UnsafeAccessor(UnsafeAccessorKind.Method, Name = "GetListOfClass2")]
[return: UnsafeAccessorType("System.Collections.Generic.List`1[[Class2, A]]")]
extern static object CallGetListOfClass2([UnsafeAccessorType("Class1, A")] object a);
```

AaronRobinsonMSFT committed:
Add support for UnsafeAccessorTypeAttribute

This commit currently only contains support for CoreCLR.

AaronRobinsonMSFT committed:
Add lookup from private assembly.

AaronRobinsonMSFT committed:
Address the collectible assembly issue.

AaronRobinsonMSFT committed:
Testing.

AaronRobinsonMSFT committed:
Offline feedback.

AaronRobinsonMSFT committed:
Update comment.

AaronRobinsonMSFT committed:
Add mono runtime support

AaronRobinsonMSFT committed:
Mono
Convert check to assert.
Include attributes when deciding to make a copy.

AaronRobinsonMSFT committed:
Remove unused local.

jkoritzinsky committed:
Add NativeAOT support

jkoritzinsky committed:
Merge branch 'unsafeaccessortype_impl' of https://github.com/AaronRobinsonMSFT/runtime into unsafeaccessortype_impl

jkoritzinsky committed:
Add linker support

AaronRobinsonMSFT committed:
Copy custom modifiers should work on N.

AaronRobinsonMSFT commented:
/cc @AndriySvyryd

AaronRobinsonMSFT committed:
Add ref return test.

AaronRobinsonMSFT committed:
Additional tests.

AaronRobinsonMSFT committed:
Remove calling convention mask.

AaronRobinsonMSFT committed:
Fix windows build break.

jkotas commented on src/coreclr/vm/siginfo.cpp:
```diff
@@ -387,6 +387,50 @@ void SigPointer::ConvertToInternalSignature(Module* pSigModule, SigTypeContext *
+    intptr_t beginExactlyOne = (intptr_t)m_ptr;
+    IfFailThrow(SkipExactlyOne());
+    intptr_t endExactlyOne = (intptr_t)m_ptr;
+    pSigBuilder->AppendBlob((const PVOID)(beginExactlyOne), endExactlyOne - beginExactlyOne);
```
These two new Copy... methods move the pointer to the end of the thing being copies.

The existing CopySignature method keep the pointer intact.

Could you please fix the existing CopySignature method to move the pointer to the end of thing being copied as well for consistency?

jkotas commented on src/tests/baseservices/compilerservices/UnsafeAccessors/PrivateLib.cs:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+
+    class Class2 { }
```
Add test coverage for generics?

Do we expect this to work for generic types that depend on the generic parameter? For example:
```
class GenericClass<T>
{
     private List<T> M1() =? new List<T>();

     private List<U> M2<U>() =? new List<U>();
}
```

AaronRobinsonMSFT replied:
> Add test coverage for generics?

Bah. Of course.

> Do we expect this to work for generic types that depend on the generic parameter?

My plan at this point was not to add support. However, I decided to look into the issues on CoreCLR and I think I can make it work. The downside is there is going to be a fair bit of complexity the consumer will need to face. I'm trying to mitigate some of that complexity at the moment.

AaronRobinsonMSFT replied:
To be clear, I've made it work without a ton of effort. The troublesome part currently is it results in the following type or declaration of accessors. I can defend the generic types in the accessors and the specification of the fully instantiated types, but if you look at `CallGenericClassIntM2String` note the use of `__Canon`. The use of `__Canon` is a bit far in my opinion and I'm currently trying to see if there is something I can do on this front. If I can get that to be `String` instead, I can defend this design - even with how ugly it is.

```csharp
class Accessors<T>
{
    [UnsafeAccessor(UnsafeAccessorKind.Constructor)]
    [return: UnsafeAccessorType("PrivateLib.GenericClass`1[[System.Int32]], PrivateLib")]
    public extern static object CreateGenericClassInt();

    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "M1")]
    [return: UnsafeAccessorType("System.Collections.Generic.List`1[[System.Int32]]")]
    public extern static object CallGenericClassIntM1([UnsafeAccessorType("PrivateLib.GenericClass`1[[System.Int32]], PrivateLib")] object a);

    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "M2")]
    [return: UnsafeAccessorType("System.Collections.Generic.List`1[[System.__Canon]]")]
    public extern static object CallGenericClassIntM2String<U>([UnsafeAccessorType("PrivateLib.GenericClass`1[[System.Int32]], PrivateLib")] object a);
}
```

jkotas replied:
Right, we do not want to leak __Canon into public surface like this.

jkotas replied:
Also, it is odd that it is `Accessors<T>` but UnsafeAccessorType is a concrete type. It is equivalent to: 
```
class Accessors<T>
{
    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "M1")]
    public extern static List<int> CallGenericClassIntM1(GenericClass<int> a);
}
```
but I believe that this use of UnsafeAccessor is unsupported. Is that correct?

AaronRobinsonMSFT replied:
> but I believe that this use of UnsafeAccessor is unsupported. Is that correct?

Correct.

Sergio0694 replied:
> "Do we know whether the motivating scenarios (EFCore, maybe CsWinRT) need generics?"

For the CsWinRT side (in case it helps), we'd primarily need non-generic support. The only scenario we have right now where we're thinking about using generics is for translating collection/dictionary interfaces between the WinRT and the .NET contract. Essentially the idea is to do something like this:

- **In WinRT.Runtime**:
```csharp
public interface IVectorViewAdapter<T>
{
    static abstract T GetAt(WindowsRuntimeObjectReference thisReference, uint index);

    // Other methods...
}

public static class IReadOnlyListMethods<T>
{
    public static T get_this<TAdapter>(WindowsRuntimeObjectReference thisReference, int index)
        where TAdapter : class, IVectorViewAdapter<T>
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index);

        T value = TAdapter.GetAt(thisReference, (uint)index);

        // Validate stuff...

        return value;
    }
}
```
- **In WinRT.Interop.dll** (the sidecar):
```csharp
static class IVectorView_stringMethods : IVectorViewAdapter<string>
{
    public static string GetAt(WindowsRuntimeObjectReference thisReference, uint index)
    {
        // Implement marshalling stub to return the value
    }
}
```
- **In any project that needs to use the `IReadOnlyList<T>` methods**:
```csharp
public static string Example(WindowsRuntimeObjectReference thisReference, int index)
{
    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod)]
    static extern string GetAt(
        [UnsafeAccessorType("IReadOnlyListMethods`1[IVectorView_stringMethods, WinRT.Interop], WinRT.Runtime")] object? _,
        WindowsRuntimeObjectReference thisReference, int index);

    return GetAt(thisReference, index);
}
```

Would this work?

Basically we'd want to invoke static methods passing some `T` that is in the sidecar .dll, which will be some generated type that will implement an "adapter" interface with all the necessary marshalling stubs for each given `T` we need.

AaronRobinsonMSFT replied:
> Would this work?

No, this contruction is impossible. We need to know `TAdapter` because it is expected in the method metadata. We can't look up which method without knowing what it should be instantiated with. We can't call a method that has method generic parameters since we don't know what which method to call.

AaronRobinsonMSFT replied:
As an example, this is what the accessor would look like:

```csharp
class Accessors<T>
{
    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name = "get_this")]
    public static extern T GetAt<TAdapter>(
        [UnsafeAccessorType("IReadOnlyListMethods`1[IVectorView_stringMethods, WinRT.Interop], WinRT.Runtime")] object? _,
        WindowsRuntimeObjectReference thisReference, int index) where U : class, ContractLib.IAdapter<T>;
}
```

We need to supply `TAdapter`.

AaronRobinsonMSFT replied:
Wait. Actually, that example doesn't make any sense. Shouldn't this be `IReadOnlyListMethods<string>`?

Sergio0694 replied:
Sorry yes, I made a typo in my example. Right, the target type would be `IReadOnlyListMethods<string>`, and the method would have `TAdapter` be the private type from the sidecar .dll, which I see you mentioned can't work. I'm trying to think of what would be the best alternative for us to use. Eg. would this work instead?

```csharp
public static string Example(WindowsRuntimeObjectReference thisReference, int index)
{
    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod)]
    static extern string GetAt(
        [UnsafeAccessorType("IReadOnlyListMethods`2[System.String, IVectorView_stringMethods, WinRT.Interop], WinRT.Runtime")] object? _,
        WindowsRuntimeObjectReference thisReference, int index);

    return GetAt(thisReference, index);
}
```

As in, if we made the type be `IReadOnlyListMethods<T, TAdapter>` 🤔

AaronRobinsonMSFT replied:
> As in, if we made the type be `IReadOnlyListMethods<T, TAdapter>` 🤔

No. There are a few things going on here. Let's enumerate them.

1) Activation of a type is done via the same machinery as `Type.GetType(string)`. This means we can "activate" basically anything as would normally be done via reflection.

2) Calling methods on types involves reflecting on the type's metadata to find the name and signature of the matching method/field. 

3) Because of the signature analysis done above (2), we must treat generics in their metadata form `ELEMENT_TYPE_VAR` and `ELEMENT_TYPE_MVAR`. This means we _need_ signatures somewhere to help with the look-up.

(1) is deceiving because people assume if we can create it, we're done. Not so fast. I mentioned this previously in https://github.com/dotnet/runtime/issues/90081#issuecomment-2811563007. 

(3) forces us to have signatures we can use for look-up. Alternatively, we could create/edit signatures, but signature manipulation is tricky and very complicated due tokens and tables and blah blah blah ..., so we went with requiring users to give us those generic signatures so it would "just work". See remarks for this in the [official doc regarding generics](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.unsafeaccessorattribute#remarks).

The `UnsafeAccessorTypeAttribute` doesn't change 1 - 3, we still need to see `ELEMENT_TYPE_VAR` and `ELEMENT_TYPE_MVAR` signatures for look-up. If the type being accessed has any unbound generic parameters those types must be visible. If the private types don't have unbound generic parameters **in the signature** we can do anything - I've **bolded** the important part.

An alternative solution for CsWinRT is to have a non-generic `static` class with N `static` functions, if the `T`s and `U`s need to be private. There are possible APIs I can forsee and we could create, but we're already at diminishing returns in my opinion and definitely not a .NET 10 feature.

jkotas replied:
> we went with requiring users to give us those generic signatures so it would "just work".

In particular, we require the unbound types in signatures to match exactly to avoid dealing with a ton of complicated ambiguous corner cases, like:
```
class G<T,U>
{
    void M(IFoo<T,U> a) { }
    void M(IFoo<U,T> a) { }
}
```

The current state PR seems to be re-introducing these problems. I understand why you have done that - `Type.GetType` does not support unbound generic arguments. But I have doubts about whether it is the right choice.

AaronRobinsonMSFT replied:
> The current state of the PR seems to be re-introducing these problems.

Right. There are still corner cases here I am thinking through. For example, I just fixed a case where I conflated "typical" with "canonical" `MethodTable`, it should have been the canonical one.

I'm fine backing out the generic support for now if you think it is a bigger can of worms than needed. EF Core, @AndriySvyryd, hasn't confirmed if they need the generic support or not yet.

Sergio0694 replied:
@AaronRobinsonMSFT thank you for the additional details! I can't say I 100% understand all the nuance there, but if you say that setup wouldn't work, I'll take your word for it. I'm thinking we could perhaps do something like this instead:

- **In WinRT.Runtime**:
```csharp
public interface IVectorViewAdapter<T>
{
    static abstract T GetAt(WindowsRuntimeObjectReference thisReference, uint index);

    // Other methods...
}

public static class IReadOnlyListMethods<T>
{
    public static T get_this<TAdapter>(WindowsRuntimeObjectReference thisReference, int index)
        where TAdapter : class, IVectorViewAdapter<T>
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index);

        T value = TAdapter.GetAt(thisReference, (uint)index);

        // Validate stuff...

        return value;
    }
}
```
- **In WinRT.Interop.dll** (the sidecar):
```csharp
static class IVectorView_stringMethods : IVectorViewAdapter<string>
{
    public static string GetAt(WindowsRuntimeObjectReference thisReference, uint index)
    {
        // Implement marshalling stub to return the value
    }
}

static class IReadOnlyList_stringMethods
{
    public static string get_this(WindowsRuntimeObjectReference thisReference, int index)
    {
        return IReadOnlyListMethods<string>.get_this<IVectorView_stringMethods>(thisReference, index);
    }
}
```
- **In any project that needs to use the `IReadOnlyList<T>` methods**:
```csharp
public static string Example(WindowsRuntimeObjectReference thisReference, int index)
{
    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod)]
    static extern string GetAt(
        [UnsafeAccessorType("IReadOnlyList_stringMethods, WinRT.Interop")] object? _,
        WindowsRuntimeObjectReference thisReference, int index);

    return GetAt(thisReference, index);
}
```

That is:
- We still keep the generic `TAdapter` setup, which allows us to build our WinRT -> .NET translation in WinRT.Runtime, instead of having to generate all of that code in each generic instantiation directly.
- For every generic instantiation we generate (eg. `IVectorView_string`), we'd also generate a non-generic class with the translation hooks for all .NET interfaces built on top of it (eg. here `IReadOnlyList_stringMethods`). Same for eg. `IVector<T>` causing `IList<T>` methods to also be generated, etc.). That is: we always generate all possibly dependent adapter hooks for each WinRT instantiation we gather.
- Code needing to use the .NET interfaces directly calls the generated non-generic adapter wrapper

I assume this is the kind of thing you also had in mind?

And I guess this also avoids the issue entirely since we no longer needs generics here 😄

AaronRobinsonMSFT replied:
> I assume this is the kind of thing you also had in mind?

Yes.

Nit: the `UnsafeAccessor` would need the `Name = "get_this"` property set.

> And I guess this also avoids the issue entirely since we no longer needs generics here 😄

Okay. I'll reach out to EF Core and cofirm their needs.

AndriySvyryd replied:
Sorry for the late reply, I am currently OOF.

EF Core wouldn't need support for generic methods, but generic types are somewhat common in our user scenarios.

AaronRobinsonMSFT replied:
> Sorry for the late reply, I am currently OOF.

@AndriySvyryd Oh, sorry.

> EF Core wouldn't need support for generic methods, but generic types are somewhat common in our user scenarios.

Are the generic types instantiated already or do you need to access them directly?

For example:

```csharp
// Would you be accessing this sort of type? Note that T needs to be supplied during access.
class A<T> { }

// - or -

// Getting the A<> instance like such? No unbound generic parameters.
class B
{
    static A<int> GetA() => new A<int>();
}
```

If you could provide a canonical example, that would help greatly.

jkotas replied:
> I'm fine backing out the generic support for now if you think it is a bigger can of worms than needed.

I think we should shoot for parity between `UnsafeAccessorType` and actual type variants. Ideally, the set of signature types that work or do not work would be exactly same between the two variants. The only exception are limitations like accessibility or unspeakable types. If we cannot get this parity, I would rather drop the generic support for now.

The problem is that Type.GetType that the implementation is based on does not support unbound generics today. Do you think that we can teach the type resolver used in this scenario to understand !0, !1, !!0, !!1, .... for unbound generic arguments? This notation is used in other places, but not in Type.GetType today. 

To implement this, we would need to add `_supportUnboundGenerics` bool to
https://github.com/dotnet/runtime/blob/41796ff0fd2a4a3f2f282563ee35cb5386c9f0f9/src/coreclr/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.CoreCLR.cs#L23, set it when we are parsing UnsafeAccessorType. and then resolve it to a generic variable here https://github.com/dotnet/runtime/blob/41796ff0fd2a4a3f2f282563ee35cb5386c9f0f9/src/coreclr/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.CoreCLR.cs#L229

AndriySvyryd replied:
Most of the times the generic type is bound, but this is also a reasonable EF scenario:

```csharp
class A<T>
{
    List<T> Ts {get; set;}
}
```

AndriySvyryd replied:
Also, for EF we know all possible values of `T` that would be needed. And in practice, it's usually just one. It would be fine if we had to create an accessor for every generic type parameter value.

I wonder whether this would also allow private  types to be used for `T`. The accessor would be called with `object`, but `UnsafeAccessorType` would be used to hardcode the private type.

AaronRobinsonMSFT replied:
Thanks @AndriySvyryd, that settles it for me.

@jkotas I'm going to remove the generic support and instead create an issue for the suggestion of recognizing the `!X` and `!!Y` syntax in the `TypeNameResolver`. There are some other APIs that could fallout if we added that support but for now it doesn't look like anyone will be blocked if we remove the generic support from `UnsafeAccessorType`. In fact it follows the intentional steps we made for `UnsafeAccessor` originally.

AaronRobinsonMSFT replied:
Hmmm... blocking this in a constructive way isn't going to be easy. Let me see what it is going to cost to implement https://github.com/dotnet/runtime/pull/114881#discussion_r2056478101.

AaronRobinsonMSFT replied:
Done.

AaronRobinsonMSFT committed:
Make SigPointer's Copy functions consistent.

jkoritzinsky committed:
Merge branch 'unsafeaccessortype_impl' of https://github.com/AaronRobinsonMSFT/runtime into unsafeaccessortype_impl

AaronRobinsonMSFT committed:
Remove duplicate throw helper.

AaronRobinsonMSFT committed:
[CoreCLR] Generic support for UnsafeAccessorType

AaronRobinsonMSFT committed:
Add tests.

jkoritzinsky committed:
Fix generics support in NativeAOT

AaronRobinsonMSFT commented on src/tests/Common/CoreCLRTestLibrary/AssertExtensions.cs:
@jkoritzinsky Did you edit this?

jkoritzinsky replied:
Yes, I ported over some more asserts from libraries because getting NativeAOT to throw COMException was annoying instead of InvalidProgramException.

AaronRobinsonMSFT committed:
Use canonical vs typical methodtable.

AaronRobinsonMSFT committed:
Update format to respect IL syntax for generics ("!N" and "!!N").

Sergio0694 commented on src/tests/baseservices/compilerservices/UnsafeAccessors/UnsafeAccessorsTests.Types.cs:
```diff
@@ -0,0 +1,300 @@
+
+using Xunit;
+
+public static class StaticClass
```
Should we also add a test with some unspeakable type that has characters that might trip the parser over? Eg. in our sidecar .dll we might have a type like this:
```
ABI.Windows.Foundation.<#Windows>TypedEventHandler`2<<#Windows>Windows-ApplicationModel-Activation-SplashScreen|object>Impl
```

I wonder if that backtick character might cause issues with the parsing logic that expects to have a generic type (whereas this type is non-generic, it just has the backtick character in its name). I'm not entirely sure `Type.GetType` properly handles it either, but perhaps even just validating that the two are consistent could make sense? 🤔

jkoritzinsky replied:
The backtick is entirely a "source language to IL compiler" convention. Nothing in ECMA requires that convention, but following it provides a mechanism for languages to support providing multiple classes with the same name with different generic parameter counts in source.

AaronRobinsonMSFT replied:
> Should we also add a test with some unspeakable type that has characters that might trip the parser over?

We follow section I.10.1 in ECMA-335 for this. If a system produces names that are inconsistent (for example, inconsistent Unicode combining) then that is a bug in the other system.

AaronRobinsonMSFT committed:
Fix build breaks.

AaronRobinsonMSFT committed:
Fix build break on linux

AaronRobinsonMSFT committed:
Enable bound generics with private types.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into unsafeaccessortype_impl

jkoritzinsky committed:
Support unbound generics in NAOT.

jkoritzinsky committed:
Fix crossgen usage of updated API

jkoritzinsky committed:
Restructure to avoid having to deal with .NET Standard-targeting ILVerify

AaronRobinsonMSFT commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -494,6 +503,121 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
+
+                        bool isMethodParameter = name.StartsWith("!!", StringComparison.Ordinal);
+
+                        if (!int.TryParse(name.AsSpan(isMethodParameter ? 2 : 1), out int index))
```
I think we should be letting exceptions flow here. I wouldn't do a try parse and instead just assume the syntax is correct.

jkoritzinsky replied:
This is in the ILC compiler, so letting this flow would crash the compiler. Right now (with the current code in this PR) a failure to parse will result in a (correctly handled) `TypeSystemException.TypeLoadException`

AaronRobinsonMSFT replied:
Does this represent a compilation error a run-time error?

jkoritzinsky replied:
Throwing a `TypeSystemException.TypeLoadException` represents a run-time error (the IL of the stub becomes just a `throw System.TypeLoadException` stub). It's the same experience as when the type being searched for is not found.

AaronRobinsonMSFT replied:
This should really be a compile time error. We know the format is wrong and we should be able to detect if the index is out of bounds. Is that possible?

AaronRobinsonMSFT replied:
Actually... if this is a build failure then tests get weird. Ugh... fine. Let's keep this as is. Thanks.

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +77,220 @@ namespace
+                cxt.Declaration /* unsafeAccessorMethod */);
+            _ASSERTE(!typeHandle.IsNull());
+
+            MethodTable* targetType = typeHandle.AsMethodTable();
```
The type handle returned by `GetTypeReferencedByCustomAttribute` can be TypeDesc. For example, consider this:
```
using System.Runtime.CompilerServices;

class Test
{
    static void Main()
    {
        object o = null;
        M2(null, ref o);
    }

    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "M")]
    static extern void M2(Test dummy, [UnsafeAccessorType("Test&")] ref object t);

    static void M(ref Test t) { }
}
```

jkotas replied:
(This fails with `Assert failure(PID 33220 [0x000081c4], Thread: 40356 [0x9da4]): !IsTypeDesc()`.)

jkotas commented:
Something that we have discussed with @AaronRobinsonMSFT offline:

> I have been mulling over the type safety. UnsafeAccessor that shipped in .NET 9 allowed skipping visibility but it did not allow violating type safety. We are allowing it to violate type safety as well now. We basically insert implicit Unsafe.As for the user. I know this is all unsafe, so it is not end of the world. I am not sure how big of a footgun we are creating here and whether it would make sense to do casts in the codegen to validate type safety, so the unsafe aspect is limited to visibility skipping only.

If we think that we are going handle missing constrains via runtime checks eventually, it may make sense to enforce type safety via runtime checks for arguments/return values as well.

Sergio0694 commented:
> "it may make sense to enforce type safety via runtime checks for arguments/return values as well"

Double checking, I assume this means no codegen difference for static methods, where `[UnsafeAccessorType]` is only used to specify the type that contains the actual method? Same for parameters/returns that don't use `[UnsafeAccessorType]` (which basically would work today already), would those also be exempt from these new possible runtime type checks?

jkotas commented:
> Double checking, I assume this means no codegen difference for static methods, where [UnsafeAccessorType] is only used to specify the type that contains the actual method? Same for parameters/returns that don't use [UnsafeAccessorType] (which basically would work today already), would those also be exempt from these new possible runtime type checks?

Right.

AaronRobinsonMSFT committed:
Handle byrefs and pointers as type name.

AaronRobinsonMSFT commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +77,220 @@ namespace
+            // We have a new type to insert. We need to update this parameter.
+            CorElementType currTypeElem;
+            IfFailThrow(origSig.GetElemType(&currTypeElem));
+            if (currTypeElem == ELEMENT_TYPE_BYREF)
```
I can do that but we are going to be opening up a lot more language in the typename support. I was trying to narrow what we should support and in that case focus on what is typically considered the type by user. The parameterized type concept is an implementation detail in my mind (i.e., byrefs and pointers). I've added tests to block them.

AaronRobinsonMSFT replied:
Well fiddlesticks. I'm looking at some other things and it would be a bit easier if the user did have to spell out the full type. Hmmmm.

jkotas replied:
Once/if we get to supporting `TypedReference`, `[UnsafeAccessorType("PrivateValueType")] TypedReference rt` would be ambiguous with this plan. It can be match either `PrivateValueType` or `PrivateValueType&`.

AaronRobinsonMSFT replied:
Yes, I've come to the same conclusion via a different scenario, but that aligns with what I've determined too.

jkoritzinsky commented:
I've started poking at the type safety checks as well.

AaronRobinsonMSFT committed:
Require UnsafeAccessorType to define full signature as a string.
Add type checks for all argument usage of UnsafeAccessorType.

AaronRobinsonMSFT committed:
Style/nits and added tests

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -477,6 +704,59 @@ namespace
+        // If the type is a byref, DUP the value so the execution stack has a
+        // byref value on it after the type check. Emit a load indirection of the byref
+        // in order to validate the type. After the typecheck, we'll pop off the object.
+        // This doesn't handle the null byref case.
```
This has type safety hole. It does not handle the case where the value that the byref is pointing to is changed between this check and use of the byref in the target method.

AaronRobinsonMSFT replied:
Can you show me an example of how this can occur? I'm not sure I understand how a byref can change in this manner. I'm interpreting this statement as follows.

```csharp

ref obj = ref ...

M(new A(), ref obj); // Is a another thread changing the object that at the ref?

extern static void M(A a, [UnsafeAccessorType("B&")] ref object o);
```

I considered the above case but there is a trade-off and I took the case that changing the value of the actual `ref` could be detected. So what I had done was create a fully typed local in the stub and pass a `ref` to that and then update it. I'm not sure which is worse. Both are detectable but the thread one seems far less common than someone expecting the `ref` to be the same.

jkotas replied:
You can hit the type safety hole on a single thread as well. Something like:
```
using System.Runtime.CompilerServices;

class Test
{
    static object o;

    static void Main()
    {
        o = "Hello";
        M(ref Unsafe.As<object, string>(ref o)); // Unsafe.As is used to simulate what's done by the UnsafeAccessor
    }

    static void M(ref string s)
    {
       o = new Test();
       Console.WriteLine(s.GetType());
    }
}
```
One option is to omit support for `ref`s for now and leave it for the future. `TypedReference`s allow passing of refs in a type-safe way.

AaronRobinsonMSFT replied:
Bah. The global state here seems a bit contrived, but I understand the example. It does seem more likely than the threading one.

> One option is to omit support for refs for now and leave it for future. TypedReferences allow passing of refs in a type-safe way.

I can do that or I can do the local approach I originally had. I will create a typed local and pass a ref to that and then update the ref on the out. Thoughts?

jkotas replied:
Is anybody asking for it right now? Leave it unsupported for now is the easy option.

Sergio0694 replied:
I can't think of a scenario where we would need `[UnsafeAccessorType]` on a byref parameter in CsWinRT off the top of my head, no. I can see how genetics might potentially be useful in some scenarios (eg. passing a derived generic type as parameter to some static method in the sidecar, or something), but can't imagine where we'd need any of those params to be byrefs. So basically dropping this (byrefs params) should be completely fine from the CsWinRT side 🙂

AaronRobinsonMSFT replied:
I should have stated this question more accurately. From the runtime perspective byrefs are expressed as `in`, `ref` and `out` in C#. None of these would be permitted without this support.

> TypedReferences allow passing of refs in a type-safe way.

@jkotas I thought a bit more about this. I don't see how a `TypedReference` is going to address the motivating example. The API is still going to accept a `ref string` and the `static` will still be `object` and assigned a `Test` instance. The `TypedReference` isn't going to stop any of those actions and the `ref` contained within the `TypedReference` instance will now point at a `Test`, but typed as a `string`. What am I missing here?

jkoritzinsky replied:
I assume that the user signature would need to take a `TypedReference` argument.

The caller wouldn't be able to pass a ref to `o` as a string-typed `TypedReference`. They'd have to construct a `TypedReference` to a `string`-typed stack location.

If the user were to construct a string-typed `TypedReference` to the location of `o`, that would require the user to write type-unsafe code (ie. using `Unsafe.As` in their code to get a `ref string` from a `ref object`).

AaronRobinsonMSFT replied:
Thanks @jkoritzinsky. Yes, that is what was intended here and I didn't understand that assumption. We discuss offline and I understand the assumption. I'm going to follow the reflection approach and make a local in the stub and pass a reference to that and then also on the out side.

jkoritzinsky replied:
Okay. I'll do something similar for the NAOT implementation.

jkoritzinsky committed:
Adjust ref handling in NAOT and add type check for values (skipping refs for now while we decide what we're going to do)

jkoritzinsky committed:
Implement byref type validation with a local for NAOT

AaronRobinsonMSFT committed:
Address type safety issue by passing
a ref to a fully typed local in the stub.

AaronRobinsonMSFT committed:
Use typed enum.

AaronRobinsonMSFT committed:
Const and nits

jkotas commented on src/coreclr/vm/siginfo.cpp:
```diff
@@ -3908,7 +3952,56 @@ MetaSig::CompareElementType(
+                    return (hInternal == hOtherType);
+                }
+                case ELEMENT_TYPE_GENERICINST:
+                case ELEMENT_TYPE_BYREF:
```
What about arrays and pointers?

For example:
```
unsafe class Test
{
    static void Main()
    {
        MProxy(null, null);
    }

    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name="M")]
    extern static void MProxy(Test t, [UnsafeAccessorType("Test[][][]")] object p);

    static void M(Test[][][] a) { }
}
```
This fails with MissingMethodException, but even though a matching method exists.

AaronRobinsonMSFT replied:
Pointers shouldn't matter at present since we require reference types only. I do need to look into arrays though.

jkotas replied:
And this:
```
using System.Runtime.CompilerServices;

#pragma warning disable CS8500

unsafe class Test
{
    static void Main()
    {
        MProxy(null, null);
    }

    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name="M")]
    extern static void MProxy(Test t, [UnsafeAccessorType("Test")] object* p);

    static void M(Test p) { }
}
```
fails with 

```
Assert failure(PID 10168 [0x000027b8], Thread: 20256 [0x4f20]): Assertion failed 'op1->TypeGet() == TYP_REF' in 'Test:MProxy(Test,ulong)' during 'Importation' (IL size 12; hash 0x3ad4888c; MinOpts)
```

AaronRobinsonMSFT commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -917,8 +917,18 @@ void CEEInfo::resolveToken(/* IN, OUT */ CORINFO_RESOLVED_TOKEN * pResolvedToken
+            {
+                // We have a TypeSpec, so we need to verify the signature is non-NULL
+                // and the typehandle has been fully instantiated.
+                if (pResolvedToken->pTypeSpec == NULL || th.ContainsGenericVariables())
```
I'm throwing a more precise error here instead of `InvalidProgramException`. I'd prefer this one the specific `TypeSpec` use case.

AaronRobinsonMSFT commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -345,7 +353,7 @@ private static bool DoesMethodMatchUnsafeAccessorDeclaration(ref GenerationConte
                 }
 
-                if (declSig.ReturnType != maybeSig.ReturnType)
+                if (declSig.ReturnType.GetTypeDefinition() != maybeSig.ReturnType.GetTypeDefinition())
```
Doesn't look like it. I will add one.

jkoritzinsky replied:
I think I added this before we added support for parsing `!0` and `!!0`. I'll try reverting these changes and see if tests pass.

AaronRobinsonMSFT replied:
Added.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.CoreCLR.cs:
```diff
@@ -228,6 +241,38 @@ internal static RuntimeType GetTypeReferencedByCustomAttribute(string typeName,
+                        // At this point we expect either another '!' and then a number or a number.
+                        bool isMethodParam = escapedTypeName[1] == '!';
+                        uint paramIndex = isMethodParam
+                            ? uint.Parse(escapedTypeName.AsSpan(2))  // Skip over "!!"
```
`NumberStyles.None` should make the parser strict.

Also, should we use TryParse and throw TypeLoadException when this is not a valid number for consistency?

AaronRobinsonMSFT replied:
Done.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into unsafeaccessortype_impl

AaronRobinsonMSFT committed:
Array support.
Error checking for generic string signature.

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,239 @@ namespace
+            // Validate the parameter resolves correctly. There is only one acceptable mapping:
+            //     ELEMENT_TYPE_OBJECT - if new type is a reference type.
+            _ASSERTE(!newTypeMaybe.IsValueType());
+            if (currHandle != TypeHandle{ g_pObjectClass })
```
> all uses of value types

Even when value types are used as generic arguments? I think it is about as much work to block it properly as it is to make it work.

AaronRobinsonMSFT replied:
Done.

jkotas commented on src/tests/baseservices/compilerservices/UnsafeAccessors/UnsafeAccessorsTests.Types.cs:
```diff
@@ -0,0 +1,477 @@
+        Assert.Equal(arg, oc);
+
+        Assert.Equal(typeof(C1[]), CallM_C1Array(tgt, Array.Empty<C1>()));
+        Assert.Equal(typeof(C1[][]), CallM_C1MDArray2(tgt, new C1[0][]));
```
This is jagged array.

MD arrays would be `C1[,]` or `C1[,,]`.

We need test coverage for both.

AaronRobinsonMSFT replied:
Oh right. I always forget that.

jkoritzinsky committed:
Remove now-incorrect generics handling and fix handling of invalid generic parameter reference syntax

jkoritzinsky committed:
Adjust handling for "ref object" parameters

jkoritzinsky committed:
Add missing case

AaronRobinsonMSFT committed:
Enforce signature validation explicitly for certain types.
Remove support for inaccessible field types due to byref to object
requirement. This was a type safety issue.
Add support for array's of inaccessible value types.

AaronRobinsonMSFT committed:
Merge branch 'unsafeaccessortype_impl' of https://github.com/AaronRobinsonMSFT/runtime into unsafeaccessortype_impl

AaronRobinsonMSFT committed:
Update to use unbox.any instead of castclass. The unbox.any
has the same semantics as castclass when the input is a reference
type.

AaronRobinsonMSFT commented:
@AndriySvyryd @Sergio0694 @jkoritzinsky FYI. Direct field access is no longer supported if the type isn't known. This would have created a type safety issue so it was removed. Instead the field type must be known _or_ accessed via property/method. If this is needed, it will need to wait until we have a usable `TypedReference`.

jkoritzinsky commented:
I'll update the NativeAOT implementation tomorrow based on recent changes 👍

AaronRobinsonMSFT commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,250 @@ namespace
+
+            // UnsafeAccessorAttribute requires the return type to be byref for any field kind.
+            // This creates a type safety issue for the runtime. Since byrefs don't support
+            // contravariance, we can't permit returning a byref to a fully typed field as a
```
```suggestion
            // covariance, we can't permit returning a byref to a fully typed field as a
```

Oops, wrong direction.

AaronRobinsonMSFT committed:
Fix mono

AaronRobinsonMSFT committed:
CoreCLR nits

jkoritzinsky committed:
Update NAOT handling

jkoritzinsky committed:
Merge branch 'unsafeaccessortype_impl' of github.com:AaronRobinsonMSFT/runtime into unsafeaccessortype_impl

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -494,6 +503,166 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
+
+                        if (isMethodParameter)
+                        {
+                            if (!method.HasInstantiation)
+                            {
+                                return null;
+                            }
+
+                            if (index >= method.Instantiation.Length)
```
```suggestion
                            if ((uint)index >= method.Instantiation.Length)
```

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -494,6 +503,166 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
+                        }
+                        else
+                        {
+                            if (!method.OwningType.HasInstantiation)
+                            {
+                                return null;
+                            }
+
+                            if (index >= method.OwningType.Instantiation.Length)
```
```suggestion
                            if ((uint)index >= method.OwningType.Instantiation.Length)
```

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.CoreCLR.cs:
```diff
@@ -21,9 +21,12 @@ internal partial struct TypeNameResolver
         private Assembly? _requestingAssembly;
         private Assembly? _topLevelAssembly;
 
+        private bool SupportUnboundGenerics { get => _unsafeAccessorMethod != IntPtr.Zero; }
```
This does not appear to be fixed

AaronRobinsonMSFT replied:
Yep. I found this last night on my other machine. Never pushed it up. Done for real now.

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -494,6 +503,166 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
+                    return SetTargetResult.Invalid;
+                }
+
+                if (replacementType.IsByRef != initialType.IsByRef)
```
Should this check be inside IsValidInitialTypeForReplacementType?

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -494,6 +503,166 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
+                if (isReturnValue && context.Kind is UnsafeAccessorKind.Field or UnsafeAccessorKind.StaticField)
+                {
+                    // We can't support UnsafeAccessorTypeAttribute on fields
+                    // today as it would create a type-safety hole.
```
This problem is not limited to fields. It applies to methods that return byrefs too.

E.g. this should be blocked for the same reason fields are blocked:
```
using System.Runtime.CompilerServices;

unsafe class Test
{
    static void Main()
    {
        ref object ro = ref MProxy(null);

        ro = "Hello";

        Console.WriteLine(g.GetType());
    }

    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name="M")]
    [return: UnsafeAccessorType("Test&")] 
    extern static ref object MProxy(Test t);

    static ref Test M() => ref g;

    static Test g;
}
```

jkoritzinsky replied:
Update NativeAOT to block ref-returning methods as well. Leaving this unresolved as CoreCLR, Mono, and tests need to be updated.

AaronRobinsonMSFT replied:
Test added.

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -504,10 +673,38 @@ private static MethodIL GenerateAccessor(ref GenerationContext context)
+                        codeStream.EmitLdArg(i);
+                        localsToRestore[i] = emit.NewLocal(targetType);
+                        codeStream.EmitLdInd(targetType);
+                        codeStream.Emit(ILOpcode.castclass, emit.NewToken(targetType));
```
VM impl does unbox.any. This should do the same

https://github.com/dotnet/runtime/pull/114881/files#diff-c5fb5a7c28b2b8c88a86c638ea93d4419b5ca1a6d8fd500f1793c9a55df758f4R781

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -538,12 +735,25 @@ private static MethodIL GenerateAccessor(ref GenerationContext context)
+                for (int i = beginIndex; i < stubArgCount; ++i)
+                {
+                    if (localsToRestore[i] != null)
+                    {
```
VM impl ships this for `in` arguments. This should do the same.

jkotas commented on src/coreclr/vm/siginfo.cpp:
```diff
@@ -3908,7 +3952,59 @@ MetaSig::CompareElementType(
+                case ELEMENT_TYPE_BYREF:
+                case ELEMENT_TYPE_ARRAY:
+                case ELEMENT_TYPE_SZARRAY:
+                case ELEMENT_TYPE_PTR:
```
Delete "Internal types can only correspond to types or value types." comment above?

jkotas commented on src/coreclr/vm/siginfo.cpp:
```diff
@@ -3908,7 +3952,59 @@ MetaSig::CompareElementType(
-                        ClassLoader::FailIfUninstDefOrRef);
+                        ClassLoader::PermitUninstDefOrRef);
+
+                    // If the return type is a generic type definition, we need to
```
I do not understand why this is needed. Is this a left over from some earlier version?

AaronRobinsonMSFT replied:
This is for the case where the type is fully instantiated in the defintion. 

```csharp
private class A
{
    private List<B> M() => default;
}
```

jkotas replied:
I would expect this case to be covered by the next `ELEMENT_TYPE_GENERICINST` below.

One of the signature is going to have `ELEMENT_TYPE_GENERICINST` and the other signature is going to have `ELEMENT_TYPE_INTERNAL`. We will load the type for `ELEMENT_TYPE_GENERICINST`, and then compare the types for exact match.

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,249 @@ namespace
+
+        // Append the new type to the signature.
+        sig.AppendElementType(ELEMENT_TYPE_INTERNAL);
+        sig.AppendPointer(th.AsPtr());
```
If we are encoding generic type with variables, we should encode generic type definition here.

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,249 @@ namespace
+            sig.AppendData(inst.GetNumArgs());
+            for (DWORD i = 0; i < inst.GetNumArgs(); i++)
+            {
+                if (!inst[i].IsGenericVariable())
```
It does sound right to skip the generic variables. They should be encoded as generic variables.

AaronRobinsonMSFT replied:
I'm not skipping generic variables. I'm only skipping if they aren't generic variables.

jkotas replied:
Still, I would expect `ELEMENT_TYPE_GENERICINST` to be always followed by the type definition and N generic arguments, irrespective of whether those arguments are generic variables or not.

AaronRobinsonMSFT replied:
Gotcha. Will update accordingly.

jkoritzinsky committed:
PR feedback

AaronRobinsonMSFT committed:
[CoreCLR]
Encode the generic definition in signature.
Encode primitive ELEMENT_TYPE in signature.

AaronRobinsonMSFT committed:
Update mono to block ref object on method return.

AaronRobinsonMSFT committed:
Use helper function.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.CoreCLR.cs:
```diff
@@ -128,14 +132,20 @@ internal static RuntimeType GetTypeReferencedByCustomAttribute(string typeName,
+            return GetTypeHelper(typeName, requestingAssembly, throwOnError, requireAssemblyQualifiedName, unsafeAccessorMethod);
         }
 
         internal static unsafe RuntimeType? GetTypeHelper(ReadOnlySpan<char> typeName, RuntimeAssembly? requestingAssembly,
             bool throwOnError, bool requireAssemblyQualifiedName)
+        {
+            return GetTypeHelper(typeName, requestingAssembly, throwOnError, requireAssemblyQualifiedName, IntPtr.Zero);
+        }
+
+        private static unsafe RuntimeType? GetTypeHelper(ReadOnlySpan<char> typeName, RuntimeAssembly? requestingAssembly,
+            bool throwOnError, bool requireAssemblyQualifiedName, IntPtr unsafeAccessorMethod)
```
```suggestion
        internal static unsafe RuntimeType? GetTypeHelper(ReadOnlySpan<char> typeName, RuntimeAssembly? requestingAssembly,
            bool throwOnError, bool requireAssemblyQualifiedName, IntPtr unsafeAccessorMethod = 0)
```
Just make it an arg with default value instead of wrapper?

AaronRobinsonMSFT replied:
I really dislike default arguments unless there is a good reason to have them. They make code reviews much harder, but if that is preferred I'll do it.

jkotas replied:
I do not have a strong opinion.

I think it is important to use named arguments when calling methods with a lot of arguments like this one. I do not think there is significant readability difference between one method with default arguments vs. multiple overloads. The calling code looks exactly same in both cases. I tend to go for the more compact variant that is the default argument variant in this case.

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -494,20 +504,254 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
+                        // if the parameter is not marked as "in".
+                        // The "sequence number" for parameters is 1-based, whereas the parameter index is 0-based.
+                        ParameterHandle paramHandle = FindParameterForSequenceNumber(reader, ref parameterEnumerator, i + 1);
+                        if (!paramHandle.IsNil)
+                        {
+                            if (!reader.GetParameter(paramHandle).Attributes.HasFlag(ParameterAttributes.In))
+                            {
+                                localsToRestore[i] = local;
+                            }
+                        }
+                        else
+                        {
```
```suggestion
                        if (paramHandle.IsNil || !reader.GetParameter(paramHandle).Attributes.HasFlag(ParameterAttributes.In))
                        {
```

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,258 @@ namespace
+            // Store the TypeHandle for the loaded type at the sequence number for the parameter.
+            cxt.TranslatedParams[seq] = { targetType, (CorParamAttr)attr };
+            attrCount++;
+        }
```
```
using System.Runtime.CompilerServices;

class Test
{
    static void Main()
    {
        MProxy<string>(null);
    }

    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name="M")]
    [return: UnsafeAccessorType("!!0")] 
    extern static object MProxy<T>(Test t);

    static T M<T>() => default;
}
```
fails with

```
Assert failure(PID 11180 [0x00002bac], Thread: 20908 [0x51ac]): Consistency check failed: AV in clr at this callstack:
------
CORECLR! MethodTable::GetInstantiation + 0xA (0x00007ffa`055f9f3a)
CORECLR! `anonymous namespace'::UpdateDeclarationSigWithTypes + 0x83D (0x00007ffa`0595886d)
CORECLR! `anonymous namespace'::ProcessUnsafeAccessorTypeAttributes + 0x9EC (0x00007ffa`05956acc)
```

AaronRobinsonMSFT replied:
Done.

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,258 @@ namespace
+                }
+                else
+                {
+                    sig.AppendElementType(ELEMENT_TYPE_INTERNAL);
```
I think this needs to run recursively to handle things like:

`List<List<!0[]>>`

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,258 @@ namespace
+    void AppendTypeToSignature(
+        SigBuilder& sig,
+        TypeHandle th,
+        Instantiation inst)
```
Does this actually need `inst` argument? Can `inst` be fetched from `th` inside the method?

AaronRobinsonMSFT replied:
I can do that. Not sure it is clearer, but I can.

AaronRobinsonMSFT commented on src/coreclr/vm/siginfo.cpp:
```diff
@@ -3908,7 +3951,52 @@ MetaSig::CompareElementType(
+                case ELEMENT_TYPE_ARRAY:
+                case ELEMENT_TYPE_SZARRAY:
+                case ELEMENT_TYPE_PTR:
+                {
+                    // Due to how SigPointer works, we need to fiddle with the signature pointer
+                    // to get the current element type back in the stream.
+                    // Since we know the size of the element type, we can just subtract one byte from the
+                    // signature pointer to get the correct value.
+                    PCCOR_SIGNATURE currInstSig;
+                    uint32_t currInstSigLen;
+                    if (Type1 == ELEMENT_TYPE_INTERNAL)
+                    {
+                        const BYTE* sigRaw = (const BYTE*)pSig2;
+                        currInstSig = (PCCOR_SIGNATURE)(sigRaw - 1);
+                        currInstSigLen = (uint32_t)((intptr_t)pEndSig2 - (intptr_t)currInstSig);
+                    }
+                    else
+                    {
+                        const BYTE* sigRaw = (const BYTE*)pSig1;
+                        currInstSig = (PCCOR_SIGNATURE)(sigRaw - 1);
+                        currInstSigLen = (uint32_t)((intptr_t)pEndSig1 - (intptr_t)currInstSig);
+                    }
+                    // Assert we are in the same state as before.
+                    _ASSERTE(*currInstSig == eOtherType);
+
+                    SigPointer inst{ currInstSig, currInstSigLen };
+                    TypeHandle hOtherType = inst.GetTypeHandleThrowing(pOtherModule, NULL);
+
+                    // SigPointer::GetTypeHandleThrowing() is non-consuming, so we need to
+                    // consume the signature to move past the current type and update the appropriate
+                    // 'in/out' parameter.
+                    IfFailThrow(inst.SkipExactlyOne());
+                    if (Type1 == ELEMENT_TYPE_INTERNAL)
+                    {
+                        pSig2 = inst.GetPtr();
+                    }
+                    else
+                    {
+                        pSig1 = inst.GetPtr();
+                    }
```
This is likely to make the emitting code complicated. Wouldn't I need to decompose the array into type _and_ the rank, size number, and bounds?

jkotas replied:
I think you need to do that anyway to handle `e.g. `List<0[,,,,]>` correctly.

AaronRobinsonMSFT replied:
Done.

AaronRobinsonMSFT committed:
Address arrays

AaronRobinsonMSFT committed:
Remove unnecessary compare logic.

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -75,16 +93,285 @@ namespace
+        }
+
+        // Extract the instantiation from parameterized type.
+        MethodTable* pMT = th.GetMethodTableOfRootTypeParam();
```
```suggestion
        MethodTable* pMT = th.GetMethodTable();
```
Param types were handled above

jkotas commented on src/coreclr/vm/unsafeaccessors.cpp:
```diff
@@ -311,7 +594,7 @@ namespace
         // Following a similar iteration pattern found in MemberLoader::FindMethod().
         // However, we are only operating on the current type not walking the type hierarchy.
-        MethodTable::IntroducedMethodIterator iter(pMT);
+        MethodTable::IntroducedMethodIterator iter(pMT, /* restrictToCanonicalTypes */ FALSE);
```
```suggestion
        MethodTable::IntroducedMethodIterator iter(pMT);
```
Is this left over from previous iterations or is this still needed?

AaronRobinsonMSFT committed:
Remove unnecessary changes.

AaronRobinsonMSFT committed:
Nits

AaronRobinsonMSFT commented on src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs:
```diff
@@ -563,6 +820,15 @@ private static MethodIL GenerateAccessorSpecificFailure(ref GenerationContext co
+            {
+                thrower = typeSysContext.GetHelperEntryPoint("ThrowHelpers", "ThrowNotSupportedException");
+            }
+            else if (isSignatureLoad)
```
Can we encode `MissingType` as a `SetTargetResult`? The extra `bool` here seems very odd.

AaronRobinsonMSFT commented on src/tools/illink/src/linker/Linker.Steps/UnsafeAccessorMarker.cs:
@sbomer Please take a look at this.

AaronRobinsonMSFT committed:
Remove specific Array branch.

jkoritzinsky committed:
PR feedback

jkoritzinsky committed:
Add UnsafeAccessorType tests

jkoritzinsky committed:
Handle null ReplacedSignatureElements

AaronRobinsonMSFT commented:
@jkoritzinsky @sbomer Looks like there is an issue with the new linker test.

```
Type `System.Runtime.CompilerServices.UnsafeAccessorTypeAttribute' should have been kept

at Mono.Linker.Tests.TestCasesRunner.AssemblyChecker.Verify() in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCasesRunner/AssemblyChecker.cs:line 89
   at Mono.Linker.Tests.TestCasesRunner.ResultChecker.Check(TrimmedTestCaseResult testResult) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCasesRunner/ResultChecker.cs:line 66
   at Mono.Linker.Tests.TestCases.All.Run(String testName) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCases/TestSuites.cs:line 154
   at Mono.Linker.Tests.TestCases.All.Reflection(String t) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCases/TestSuites.cs:line 79
   at InvokeStub_All.Reflection(Object, Span`1)
   at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

jkoritzinsky commented:
I ran the tests for ILLink but not ILC. I'll fix that.

jkoritzinsky committed:
Fix NAOT trimming test

AaronRobinsonMSFT committed:
Update API documentation.

jkoritzinsky commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/UnsafeAccessorTypeAttribute.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <ul>
+        ///   <li>References should be typed as <lang>object</lang>.</li>
+        ///   <li>Arrays should be typed as <lang>object</lang>.</li>
+        ///   <li>Byref arguments should be typed with <lang>in</lang>, <lang>ref</lang>, or <lang>out</lang> to <lang>object</lang>.</li>
+        ///   <li>Pointers should be typed as <lang>void*</lang>.</li>
```
```suggestion
        ///   <li>Pointers should be typed as <lang>void*</lang>.</li>
        ///   <li>Byref arguments to reference types or arrays should be typed with <lang>in</lang>, <lang>ref</lang>, or <lang>out</lang> to <lang>object</lang>.</li>
        ///   <li>Byref arguments to pointer types should be typed with <lang>in</lang>, <lang>ref</lang>, or <lang>out</lang> to <lang>void*</lang>.</li>
```

AaronRobinsonMSFT committed:
Update comments.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/UnsafeAccessorTypeAttribute.cs:
```diff
@@ -0,0 +1,54 @@
+        ///
+        /// <ul>
+        ///   <li>References should be typed as <lang>object</lang>.</li>
+        ///   <li>Arrays should be typed as <lang>object</lang>.</li>
```
```suggestion
```
Arrays are reference types. I do not think you need to mention them explicitly.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/UnsafeAccessorTypeAttribute.cs:
```diff
@@ -0,0 +1,54 @@
+        ///   <li>References should be typed as <lang>object</lang>.</li>
+        ///   <li>Arrays should be typed as <lang>object</lang>.</li>
+        ///   <li>Byref arguments should be typed with <lang>in</lang>, <lang>ref</lang>, or <lang>out</lang> to <lang>object</lang>.</li>
+        ///   <li>Pointers should be typed as <lang>void*</lang>.</li>
```
```suggestion
        ///   <li>Unmanaged pointers should be typed as <lang>void*</lang>.</li>
```
If you are using ECMA-335 terminology, this should be "unmanaged pointers".

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/UnsafeAccessorTypeAttribute.cs:
```diff
@@ -0,0 +1,54 @@
+        /// <remarks>
+        /// <paramref name="typeName"/> is expected to follow the same rules as if it were being
+        /// passed to <see name="Type.GetType(String)"/>. When unbound generics are involved they
+        /// should follow the IL syntax of referencing an ELEMENT_TYPE_VAR or ELEMENT_TYPE_MVAR using
```
```suggestion
        /// should follow the IL syntax of referencing a type or method generic variables using
```

AaronRobinsonMSFT committed:
Feedback

jkoritzinsky commented:
/ba-g deadletter
